test/cpp/end2end/test_service_impl.cc

```diff
@@ -444,16 +445,20 @@ experimental::ServerUnaryReactor* CallbackTestServiceImpl::Echo(
     }
 
     void StartRpc() {
-      if (req_->has_param() && req_->param().server_sleep_us() > 0) {
-        // Set an alarm for that much time
-        alarm_.experimental().Set(
-            gpr_time_add(gpr_now(GPR_CLOCK_MONOTONIC),
-                         gpr_time_from_micros(req_->param().server_sleep_us(),
-                                              GPR_TIMESPAN)),
-            [this](bool ok) { NonDelayed(ok); });
-      } else {
-        NonDelayed(true);
+      {
+        std::lock_guard<std::mutex> l(alarm_mu_);
+        if (req_->has_param() && req_->param().server_sleep_us() > 0 &&
+            !dont_alarm_anymore_) {
```

Comment:
If the server is supposed to sleep for a non-zero amount of time, but we're not supposed to alarm, doesn't that indicate a bug? Why aren't we doing a CHECK(!don't_alarm_anymore) instead?

Reply:
This situation would arise in the future if we were to add a delayed-server test that also had a client-side cancellation involved. It doesn't arise for any of our current tests. I suppose I could add such a test to make this entire change more meaningful since the purpose was to reduce the fragility of the logic to adding new tests.

Reply:
Recap from our offline discussion:
The concept we're trying to capture in this if check is "If a cancellation has already occurred, there is no need to sleep, so just finish immediately." This concept doesn't come across clearly because we're using dont_alarm_anymore_, an implementation detail, rather than on_cancel_invoked_, which captures the concept of cancellation. The confusion can be partially resolved by getting rid of dont_alarm_anymore_, which I'll describe in a comment elsewhere.

I actually think it would be preferable to not check this variable in the if statement. Even if we don't *need* to sleep, it's simpler for the server to sleep as the test says rather than to *not* sleep under a particular timing just for the sake of saving some test time. I can imagine someone debugging why the server seems to sleep in some cases and not in others despite being told to sleep.